registration/views.py

```diff
@@ -56,30 +66,135 @@ def user_data_view(request):
+        if 'print' in request.POST:
+            assert self.profile
+            return generate_registration_form(self.user, self.profile)
+        else:
+            assert 'manual-form' in request.POST
+            return http.FileResponse(
+                open('files/Nutzerkartei_Anmeldung_2017.pdf', 'rb'),
+                filename=('application_form.pdf')
+            )
+
+
+@method_decorator(login_required, name='setup')
+class EditProfileView(generic.View):
+    """View so the user can edit his/her profile."""
+
+    template_name = 'registration/edit_profile.html'
+
+    def setup(self, request, *args, **kwargs):
+        """Initialize attributes."""
+        self.user, self.profile = _get_user_and_profile(request)
+        self.initial_data = _get_initial_data(self.user, self.profile)
+
+        return super(EditProfileView, self).setup(request, *args, **kwargs)
+
+    def get(self, request):
+        """
+        Handle get requests.
+
+        Create form with user data and writeable fields depending on the users
+        verification.
+        """
+        if not self.profile:
+            return _no_profile_error(request)
+
+        form = UserDataForm(self.initial_data)
+
+        ALWAYS_WRITABLE = ['phone_number', 'mobile_number']
+        # if verified only ALWAY_WRITABLE fields are writeable
+        if self.profile.verified:
+            for field in form.fields.keys():
+                if field not in ALWAYS_WRITABLE:
+                    form.fields[field].widget.attrs['readonly'] = True
+
+        return render(
+            request,
+            self.template_name,
+            {'form': form, 'user': self.user}
         )
 
+    def post(self, request):
+        """Submit changes of the users data."""
+        assert 'submit' in request.POST
+        assert self.profile
+
+        form = UserDataForm(request.POST, initial=self.initial_data)
+
+        if not form.is_valid():
+            return _validation_errors(request, self.template_name, form)
+
+        cleaned_data = form.cleaned_data
+
+        for field in form.changed_data:
+            if hasattr(self.profile, field):
+                setattr(self.profile, field, cleaned_data[field])
+            else:
+                # It can only be the email
+                assert field == 'email'
+
+                # Does the email already belongs to another user?
+                email = cleaned_data['email']
+                used = User.objects.filter(email=email)
+                if used and used[0].id != self.user.id:
+                    messages.error(
+                        request, f'The e-mail address {email} already exists.')
+                    return render(request, self.template_name, {"form": form})
+
+                self.user.email = email
+
+        self.user.save()
+        self.profile.save()
+
+        messages.success(request, _('Your profile was successfully updated.'))
+        return render(request, self.template_name, {"form": form})
```

Comment:
As the form is rendered like this, the fields are not set as read-only in case of a verified profile. Maybe we redirect to the GET view or we have to set the fields to read-only here too.

Reply:
Using the get function worked ([4e740ff](https://github.com/gocept/ok_tools/pull/23/commits/4e740ff3abcbd7c2dfffd55ca5b180f343e39542))